[feature/lecture/recommendation] ✨ ALS 기반 강의 추천 시스템(비동기 작업 포함) 추가

209512 opened the pull request:
## ✅ PR 요약  
- 관련 이슈 번호: #REQ-LECT-003  
- 작업 요약: ALS 기반 강의 추천 시스템 구현 (Full/Partial Training, 비동기 작업, 테스트 포함)  
  
## 📄 상세 내용  
  
### 1. 추천 시스템
  
- [x] **constants.py**: 시스템 상수 및 설정 정의  
  - ALS 하이퍼파라미터 (factors=50, regularization=0.01, iterations=15)  
  - 상호작용 가중치 (북마크 3.0, 스터디 2.0, 검색 1.0)  
  - 시간 감쇠 설정 (반감기 7일)  
  - Redis 캐시 키 및 TTL 설정  
  - 후처리 보너스 계수 (카테고리 매칭 2.5, 평점 1.5)  
  
- [x] **data_loader.py**: 사용자-강의 상호작용 행렬 구축  
  - 북마크, 검색 로그, 스터디 참여 데이터 통합  
  - 시간 감쇠 적용: `0.5^(경과일수/반감기)`  
  - COO/CSR sparse matrix 변환  
  - L1 정규화로 사용자별 활동량 차이 제거  
  - 신규 사용자/강의 조기 검증  
  
- [x] **model_trainer.py**: ALS 모델 학습 및 관리  
  - **Full Training**: 전체 데이터로 모델 학습 (초기 학습, Partial 실패 시)  
  - **Partial Training**: 증분 학습 (신규 데이터만, 수 초 소요)  
  - Redis 락 기반 동시성 제어 (Exponential backoff, 최대 5회 재시도)  
  - Atomic 파일 저장 (joblib + xz 압축 + 백업/롤백)  
  - 캐시 동기화 (모델, 매핑, 행렬)  
  - 신규 사용자 감지 시 Full Training 자동 폴백  
  
- [x] **recommender.py**: 추천 서비스 메인 로직  
  - ALS 모델 기반 협업 필터링  
  - 다단계 폴백 전략: ALS → 카테고리 기반 → 인기 강의  
  - 후처리 스코어링: `normalized_als_score + rating_bonus + category_bonus`  
  - Redis 캐싱 (모델, 메타데이터, 인기 강의)  
  - Redis 헬스체크 및 자동 재연결 (1시간 주기, 실패 시 60초 주기)  
  - Redis pipeline으로 메타데이터 일괄 조회 (N+1 방지)  
  
### 2. API 엔드포인트  
  
- [x] **RecommendationView**: 사용자 맞춤 추천 강의 조회 API  
  - 인증 필요 (`IsAuthenticated`)  
  - 추천 개수 제한 (3개)  
  - 북마크 제외 로직  
  - 응답 형식: `{'detail': '...', 'data': {'recommendations': [...]}}`  
  
### 3. 비동기 작업 (Celery)  
  
- [x] **tasks.py**: 모델 학습 비동기 작업  
  - `train_full_model_task`: 전체 모델 학습 (타임아웃 1시간, 재시도 3회)  
  - `partial_fit_model_task`: 증분 학습 (타임아웃 30분, 재시도 3회)  
  - Exponential backoff 재시도 전략  
- [x] **add_celery_tasks.py**:  주기 설정
  - full training: 매주 일요일 오전 2시 0분 실행 설정 (day_of_week="0").
  - partial-fit-daily: 매일 오전 3시 0분 실행 설정 (hour="3", minute="0").

  
### 4. 캐시 무효화  
  
- [x] **signals.py**: 강의 정보 변경 시 캐시 자동 무효화  
  - `post_save`, `post_delete` 시그널 처리  
  - 강의별 메타데이터 캐시 삭제  
  - 인기 강의 캐시 삭제  
  
---
## 시스템 구조

```
HTTP Request
↓
RecommendationView (API)
↓
RecommendationService
├─ ALS 모델 기반 추천
├─ 카테고리 폴백
└─ 인기 강의 폴백
↓
DataLoader (행렬 구축)
├─ 북마크 데이터
├─ 검색 로그
└─ 스터디 참여
↓
ModelTrainer (학습/저장)
├─ Full Training (Celery)
└─ Partial Training (Celery)
↓
Redis Cache + File Storage
```
---
  
## 📝 기타 참고 사항  
  
### 로깅 태그
- `[CACHE]`: 캐시 관련 동작  
- `[REC]`: 추천 로직  
- `[RECSCORE]`: 스코어링 상세 (DEBUG)  
- `[FALLBACK]`: 폴백 전략 실행  
- `[ALS]`: 모델 학습/저장/로드  
- `[MATRIX]`: 행렬 구축  
- `[DB]`: 데이터베이스 쿼리  
- `[CELERY]`: Celery 비동기 작업  
  
### 주요 설정  
- 모델 파일 저장 경로: `settings.MODEL_STORAGE_PATH`  
- Redis 캐시 TTL: 모델(7일), 메타데이터(1일), 인기 강의(1시간)  
- ALS 하이퍼파라미터: factors=50, regularization=0.01, iterations=15  
- 시간 감쇠 반감기: 7일  
- Redis 락 타임아웃: 60초  
- Celery 타임아웃: Full Training(1시간), Partial Training(30분)  
  
### 주의사항  
- `MODEL_VERSION` 변경 시 캐시 자동 무효화  
- `implicit` 라이브러리 버전 호환성 확인 필요  
- `partial_fit`은 shape 변경 불가 (신규 사용자/강의 시 Full Training 필수)  
- 통합(E2E) 테스트는 실제 모델 학습을 포함하므로 실행 시간 소요 (0.2~0.5초)  
  
## 🧪 PR Checklist  
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

209512 committed:
💡 chore(lecture): rebase conflicts

# Conflicts:
#	poetry.lock
#	pyproject.toml

# Conflicts:
#	apps/lecture/urls/__init__.py
#	poetry.lock
#	pyproject.toml

# Conflicts:
#	poetry.lock

209512 committed:
🐛 fix(lecture): 추천 강의 ID 조회 시 KeyError 방지 및 폴백 처리 추가

테스트 코드 작성 중 KeyError 발견.
- 추천 결과에서 잘못된 인덱스 발생을 방지하기 위해 추천 인덱스 존재 여부 체크 후 리스트에 추가.
- 추천 결과 없을 경우 인기 강의로 안전하게 폴백.

209512 committed:
✅ test(lecture): recommendation 통합 테스트 코드 추가

209512 committed:
🐛 fix(lecture): RecommendationService 지연 초기화

ci 에러
- Django 마이그레이션 실행 시점에 데이터베이스에 테이블이 아직 생성되지 않은 상태에서
- 코드가 초기화(임포트) 과정 중에 해당 테이블(lecture_categories)에 접근하려고 시도하여 발생한 django.db.utils.ProgrammingError: relation "lecture_categories" does not exist 에러
- 즉, 마이그레이션이 완료되기 전 DB 스키마와 코드 간 불일치로 생긴 테이블 미존재 오류

-> 서비스 클래스의 DB접근 코드를 지연 초기화(lazy init) 방식으로 변경
-> 뷰에서 서비스 인스턴스 생성 시점을 요청 시점으로 변경

2joonkim commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,252 @@
+
+        # 2. 검색어 기반 점수 누적: 관심사 파악
+        search_keywords = LectureSearchLog.objects.filter(user_id=user_id).values_list("keyword", flat=True)
+        for keyword in search_keywords:
```
이 부분 심각한 n+1 오류가 보입니다.
검색 한번당 db쿼리를 1번씩 쏴야하는데 개선이 필요해 보입니다

209512 committed:
🐛 fix(lecture): N+1 개선

209512 committed:
🐛 fix(lecture): 추천 서비스 전반 쿼리 최적화

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+            shape=(len(users), len(lectures)),
+        )
+
+        return user_item_matrix, user_idx_map, lecture_idx_map, users, lectures
```
여기까지는 Model Training을 위한 행렬 데이터 준비를 위한 코드인데 별도의 데이터 로더 혹은 전처리 클래스로 두시면 좋을 것 같아요
지금은 매번 추천 서비스를 호출할 때마다 데이터를 수집하고 행렬 매트릭스를 생성하고, 모델훈련 하는과정을 반복하고 있는데 
```
1. 특정 시간마다 전체 모델 학습을 새로 진행하는 배치 작업 처리하여 훈련된 모델을 joblib으로 로컬 저장하기
2. 새로운 데이터가 추가될 때 마다 로컬에 저장된 훈련 모델을 불러와서 fit_partitial 메서드를 통해 추가된 데이터를 통해 fit_partitial 메서드를 통해 점진적 학습시키고 모델을 새로 joblib으로 로컬에 저장하기
(해당 작업도 처리시간이 길 수 있으니 Celery 같은 비동기 작업 처리 요망)
3. 추천 요청이 있을 때마다 해당 저장된 모델을 활용하여 항목 추천 리스트 반환하기
```
이렇게 로직흐름을 가져가보면 좋을 것 같습니다.

implicit 라이브러리의 AlternatingLeastSquares 모델은 기본적으로는 batch 학습용으로 설계되어 있지만 fit_partitial 메서드를 통해 점진적 학습도 가능하기 때문에 좋은 선택인 것 같습니다.

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+        # 모델 훈련: Item x User 행렬 (matrix.T)을 CSR 형식으로 변환하여 효율적으로 훈련
+        model.fit(matrix.T.tocsr())
+
+        return matrix, model, u_to_i, l_to_i, users, lectures
```
이 트레이닝 함수는 별도의 LectureRecommanderModelTrainer 로 분리하시고 분리된 클래스내에서 위의 피드백 적용해보시면 좋겠습니다.

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+}
+
+
+class RecommendationService:
```
```
services/
- data_loader.py      # Django ORM 결과를 Sparse Matrix 변환하는 담당
- model_trainer.py    # ALS 학습 및 모델 저장 및 불러오기
- recommender.py      # 추천 로직 및 폴백 처리
```
이렇게 분리해보면 좋을 것 같아요

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+        "search": 1.0,  # 검색: 낮은 강도의 관심 표현
+        "study_participation": 2.0,  # 스터디 참여: 높은 몰입 신호
+        "category_match": 2.5,  # 카테고리 일치: 장기 선호도 반영
+        "review_rating": 1.5,  # 리뷰 평점: 품질 만족도 반영
```
각 가중치 점수를 하드코딩하지말고 Constants로 관리하시면 추후에 가중치 변경해야하는 상황이나 유지보수에 용이합니다.

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+            regularization=0.1,
+            iterations=15,
+            calculate_training_loss=False,
+            use_gpu=False,
```
모델 훈련에 필요한 값들도 값을 조율해가면서 모델 훈련을 진행하고 결과를 비교하는 경우가 있을텐데 하드코딩하지말고 Constants로 관리하시면 용이해요

Meoyoug replied:
또는 DataClass 활용해보기

Meoyoug commented on apps/lecture/services/recommendation_service.py:
```diff
@@ -0,0 +1,260 @@
+            self._lecture_category_map = self._load_lecture_category_map()
+        return self._lecture_category_map
+
+    def aggregate_user_interactions(self, user_id: int) -> Dict[int, float]:
```
여기서 매 유저별 aggregate_user_interactions(user_id) 호출 하면 각 항목 집계를 위해서 DB를 5번 이상 호출하는 것 같아요.
O(user_count) * (query_count) 형태로 쿼리 개수가 늘어날 수록 해당함수가 점점 느려질 겁니다.

가능한부분에 대해서는 DB 호출 후에 가중치 점수 계산을 하기보다 DB 쿼리 호출하면서 집계 쿼리 사용해서 미리 계산해서 들고오시는게 좋아요

209512 committed:
🐛 fix(lecture): 추천 서비스 피드백 반영

- 로직 분산: 기존 service.py를 data_loader, model_trainer, recommender 파일로 분리
- constants.py 생성: 모델 관련 상수
- 집계 쿼리 사용
- 모델 저장 경로: settings.py에 MODEL_STORAGE_PATH 정의 및 환경 변수 우선순위 적용

209512 committed:
✅ test(lecture): recommendation 통합 테스트 코드 변경

209512 committed:
🎨 style(lecture): code formatting 수정

Meoyoug commented on apps/lecture/services/model_trainer.py:
```diff
@@ -0,0 +1,90 @@
+
+            return model, u_to_i, l_to_i, users, lectures
+        except Exception as e:
+            print(f"Error loading model or mappings: {e}")
```
print문 대신 로깅 활용해주세요

Meoyoug commented on apps/lecture/tests/test_recommendation_integration.py:
```diff
@@ -0,0 +1,172 @@
+        recs: Union[List[Dict[str, Any]], None] = response.data["data"]["recommendations"]
+        self.assertIsInstance(recs, list)
+
+        if recs:
```
여기서 if 절 제외시켜주세요

data 키값의 밸류로 recommendations 키, 밸류가 존재하는 구성에 대한 테스트 인거잖아요?
테스트 함수 내에서 분기는 없애거나 최소화하셔야합니다.

Meoyoug commented on pyproject.toml:
```diff
@@ -24,6 +24,12 @@ dependencies = [
+    "pandas (>=2.3.3,<3.0.0)",
+    "implicit (>=0.7.2,<0.8.0)",
+    "pandas-stubs (>=2.3.2.250926,<3.0.0.0)",
+    "scipy-stubs (>=1.16.2.4,<2.0.0.0)",
```
stubs 들은 dev dependency로 추가부탁드려용

Meoyoug commented on apps/lecture/services/data_loader.py:
```diff
@@ -0,0 +1,170 @@
+            lecture_category_map.setdefault(lec_id, set()).add(cat_id)
+        return lecture_category_map
+
+    def get_lecture_category_map(self) -> Dict[int, set[int]]:
```
이 매핑은 redis를 활용해서 캐시해주세요

209512 committed:
🐛 fix(lecture): N+1 방지

Meoyoug commented on apps/lecture/services/recommender.py:
```diff
@@ -0,0 +1,118 @@
+        self._user_to_idx: Optional[Dict[int, int]] = None
+        self._lecture_to_idx: Optional[Dict[int, int]] = None
+        self._lecture_idx_to_id: Optional[Dict[int, int]] = None
+        self._user_items_matrix: Optional[csr_matrix] = None
```
다중 워커 환경이라면 각 워커 프로세스마다 메모리에 모델 및 매핑정보를 캐시하기 때문에 메모리가 부족해질 수 있어요 redis를 사용해서 캐싱하는 걸 알아보시면 좋을 것 같아요

209512 committed:
🐛 fix(lecture): Redis 캐싱 적용(피드백)

1. Redis 캐싱 적용
- data_loader: def get_lecture_category_map
pyproject.toml: stubs들 dev dependency로 추가
- ALS 모델, 매핑 정보(인덱스) Redis로 변경
2. Testcode 내 if 절 제거
3. 스터디 참여 로직의 예외 처리 구체화
4. 검색 로그를 최근 30일 데이터만 사용하도록 필터링

209512 committed:
🐛 fix(lecture): joblib 사용, groupmember 추가

1. ALS 모델 저장 및 로드에 implicit 내장 메서드 대신 joblib.dump()/joblib.load() 사용
2. 런타임 시 GPU(CUDA) 환경을 자동 탐지하고, 초기화에 실패할 경우 CPU ALS로 자동 폴백(Fallback) 추가
- constants.py: ALS_Hyperparameters의 use_gpu 제거
3. 이전 StudyLecture 모델만으로는 study_group_id와 lecture_id 쌍만 확인 가능
-> 스터디 그룹의 구성원 알 수 없음
-> 구성원의 상호작용 행렬의 위치에 점수 추가 불가
=> GroupMember 모델 추가

209512 committed:
✨ feat(lecture): logging, 쿼리 최적화, 추천결과 부족 시 인기 강의 폴백

209512 committed:
✅ test(lecture): recommendation 테스트 코드 변경

209512 committed:
✨ feat(lecture): fit_partial(점진적 학습) 추가

DataLoader 수정 (data_loader.py):
- `_get_recent_interactions(since)` 메서드 추가
    -`since` 시점 이후에 발생하거나 업데이트된 북마크, 스터디 참여, 검색 로그 등 신규 상호작용 데이터만 조회.
 - `build_partial_user_item_matrix` 메서드 추가:
     - 기존 사용자/강의 매핑을 유지한 채 신규 상호작용으로 구성된 희소 행렬(Partial COO Matrix) 생성.

ModelTrainer 수정 (model_trainer.py):
- `train_and_save_partial_model(since: timedelta)` 메서드 추가.
- 기존 모델 및 매핑을 로드한 후, 로드된 부분 행렬을 `model.fit_partial()`에 전달
-> 기존 학습 결과를 유지하며 모델을 빠르게 업데이트.
- 학습 데이터가 없을 경우 로깅 후 학습을 건너뛰도록 처리.

209512 committed:
🐛 fix(lecture): partial_fit_users (점진적 학습) 개선

점진적 학습 개선: 새로운 강의(아이템) 유입 시 모델의 인덱스 매핑이 깨지는 것을 방지하기 위해, 신규 강의가 발견되면 전체 재학습 유도
DataLoader.build_partial_user_item_matrix의 반환 값이 업데이트 대상 사용자 ID 리스트와 부분 행렬, 신규 강의 유무 플래그로 변경되어 ModelTrainer에서 partial_fit_users 메서드를 사용하여 사용자 요인만 효율적으로 업데이트하도록 변경
날짜 처리 수정: datetime.now() -> Django의 timezone.now()
로깅 레벨 조정: error -> warning

209512 committed:
🐛 fix(lecture): try-except 개선

209512 committed:
✨ feat(lecture): 추천 시스템 안정성 및 효율성 개선 (모델 버전 자동화, 안전 저장, 점진 학습 강화)

- Model Versioning: 현재 날짜 기반 버전 자동 생성 및 모든 Redis 캐시 키에 버전 포함
- Cache Management: 모델 재학습/배포 시 구버전 캐시 자동 무효화 추가
- Constants Refactor: 상호작용 가중치, 아이템 피처 가중치 분리; ALS 감쇠, 정규화 등 후처리 정책 상수 추가
- ModelTrainer: 모델 파일 저장 시 안전 저장 (백업/복구) 로직 도입
- ModelTrainer: 학습 성공 후 관련 Redis 캐시 키 일괄 삭제 로직 추가
- DataLoader: 점진 학습용 데이터 로드 (`_get_recent_interactions`) 로직 개선 및 상호작용 피처 통합 강화
- ModelTrainer: 점진 학습 불가 시 (신규 아이템/사용자 포함) 자동 전체 재학습 폴백 추가

209512 committed:
✨ feat(lecture): ALS 추천 후처리(Reranking) 도입, 콜드 스타트 개선 및 API 최적화

- Reranking 도입: `RecommendationService`에 `_get_post_processed_ranking` 메서드 추가
- Reranking Logic: ALS 점수 감쇠/정규화 후, 평점 및 선호 카테고리 보너스를 적용해 최종 순위 재정렬
- Cold Start/Fallback 개선: 모델 실패 및 신규 사용자 대상 '선호 카테고리 기반 랜덤 샘플링' 폴백 추가
- Model Load Stability: Redis 상태 점검, 복구 주기 설정 및 락 기반 모델 로드 로직 도입 (`_ensure_model_loaded`)
- N+1 방지: `RecommendationView`에서 북마크 ID를 미리 조회하여 Serializer Context로 전달
- API 응답 구조를 강의 목록(`serializer.data`)만 반환하도록 단순화
- 함수명 변경: `recommend_lectures_for_user` -> `recommend_lectures`

209512 committed:
♻️ refactor(lecture): ALS 추천 시스템 핵심 모듈 구조 개선 및 최적화

- DataLoader 개선: 희소 행렬 정규화 최적화, Partial Fit 사용자/아이템 처리 강화
- ModelTrainer 구현: atomic 저장, 백업/복구 로직, Redis 캐시 동기화, 락 기반 동시성 제어 적용
- RecommendationService 개선: Redis 상태 점검, 안전한 캐시 get/set, 모델 로드 및 캐시 폴백 전략 강화
- ALS 추천 결과 후처리 고도화: 점수 감쇠 및 정규화, 평점과 카테고리 보너스 적용
- 콜드 스타트 대응: 인기 강의 및 사용자 선호 카테고리 기반 폴백 추천 추가

209512 committed:
🎨 style(lecture): code formatting

209512 committed:
✅ test(lecture): recommendation api 테스트 작성

209512 committed:
🚚 build(lecture): poetry.lock 수정

209512 committed:
🐛 fix(lecture): 추천 기능 논리 오류 개선- 빈 행렬, 시간 처리

테스트 코드 작성 중 발견, 이하 조치.

### 주요 변경사항

1. constants.py: 모델 버전 관리 개선
- MODEL_VERSION을 환경변수 기반으로 변경 (ALS_MODEL_VERSION)
- 테스트 환경 지원: TEST_MODEL_VERSION 환경변수 추가
- datetime.now() -> timezone.now() 변경: timezone-aware datetime 사용
- 분산 시스템에서 캐시 키 불일치 문제 해결
- 자정 전후 자동 버전 변경으로 인한 캐시 무효화 문제 해결

2. data_loader.py: 데이터 로더 시간 처리 및 캐싱 개선
- _get_current_time() 메서드 추가로 일관된 시간 기준 제공
- 모든 감쇠 계산에 reference_time 파라미터 전달
- 검색 로그 기간 제한 로직 개선 (effective_start = max(last_trained_at, cutoff))
- 검색 빈도 반영: 동일 키워드 중복 검색 시 각 검색마다 점수 부여
- 스터디 참여 로직 개선: 그룹 가입 시점 검증 추가
- 캐시 키 타입 안전성 확보: bytes, 문자열, 정수 모두 처리
- 빈 행렬 처리 로직 명확화: Full Fit과 Partial Fit 구분

3. model_trainer.py: 모델 학습 로직 개선 및 최적화
- datetime.utcnow() -> timezone.now() 변경
- Partial Fit 시 행렬 Shape 불일치 처리 로직 개선
  - 신규 강의 유입 감지 및 Full Training 폴백
  - 기존 강의 일부 상호작용 시 행렬 확장 로직 개선
  - NumPy 벡터화로 행렬 확장 성능 최적화 (루프 제거)
- last_trained_at timezone-aware 변환 추가
- return_format 파라미터 추가로 불필요한 행렬 변환 제거

4. recommender.py: 추천 서비스 캐싱 및 Fallback 로직 개선
- 메타데이터 캐싱 직렬화 방식 통일 (JSON -> pickle)
- Redis pipeline 사용으로 메타데이터 조회 성능 최적화 (N번 -> 1번 네트워크 호출)
- 인기 강의 캐싱 최적화: DB 쿼리 1회 실행, 메모리에서 북마크 필터링
- Fallback 로직 일관성 확보: ALS 실패 시 Category Fallback으로 통일
- Redis 연결 상태 확인 개선: 불필요한 로그 출력 제거
- last_trained_at timezone-aware 변환 추가

### 발견(해결) 문제점

버전 관리 및 시간 처리
- MODEL_VERSION 동적 생성으로 인한 캐시 불일치
- Timezone naive/aware datetime 혼용
- 시간 감쇠 계산 기준 시각 불일치
- 테스트 환경에서 예측 불가능한 버전 생성

데이터 로딩 및 처리
- Partial Fit 시 검색 로그 기간 제한 데이터 누락
- 검색 빈도 무시 문제
- 스터디 참여 시점 검증 누락
- 캐시 키 타입 불일치 위험 (bytes 타입 미처리)
- 빈 행렬 처리 로직 모호함

모델 학습
- Partial Fit 행렬 Shape 불일치
- 행렬 확장 로직 비효율 (루프 기반)

추천 서비스
- 메타데이터 캐싱 직렬화 불일치
- 인기 강의 조회 중복 쿼리
- 메타데이터 조회 성능 저하 (N번 네트워크 호출)

### 성능 개선

- 행렬 확장 로직: NumPy 벡터화로 대량 강의 처리 시 성능 향상
- 메타데이터 캐싱: Redis pipeline으로 네트워크 호출 횟수 대폭 감소
- 인기 강의 조회: 중복 DB 쿼리 제거로 응답 시간 단축

209512 committed:
🐛 fix(lecture): 행렬 Shape 불일치 오류. 행렬 확장 로직-> NumPy 벡터화

recommender.py:
- Redis 연결 실패 시 Django 캐시로 자동 폴백 개선 (_cache_get_safe, _cache_set_safe).
- Redis Pipeline을 사용하여 강의 메타데이터 대량 조회 시 네트워크 호출 횟수 감소 개선
- last_trained_at 등 시간 관련 데이터의 Timezone-aware 변환 로직 누락 개선
- ALS 실패, 콜드 스타트 시 다계층 폴백 전략 (ALS -> 카테고리 -> 인기 강의) 개선.

model_trainer.py, data_loader.py:
- Partial Fit 시 신규 사용자 유입 또는 신규 강의 유입 시 Full Training으로 폴백 로직 보완: 행렬 Shape 불일치 오류 개선.
- 모델 저장 시 Atomic Save 및 Rollback 보완,
- Redis 락 로직을 최종 점검 및 보완.
- 검색 로그 빈도 반영 및 스터디 참여 가입 시점 검증 로직 개선

.gitignore: 모델 저장 경로 추가

209512 committed:
✅ test(lecture): data_loader integration, model_trainer integration, recommender integration, recommendation api, recommendation integration 테스트 추가 및 수정

209512 committed:
🚚 build(lecture): poetry lock

209512 committed:
🐛 fix(lecture): 누락 수정: 추천-북마크된 강의 제외

209512 committed:
✅ test(lecture): recommendation 관련 테스트들/ 일부 base_lecture test 상속으로 변경, unit 테스트 추가

209512 committed:
💡 chore(lecture): recommendation services 주석 정리

209512 committed:
💡 chore(lecture): 오타 수정

setting.DEBUG 분기 제거 및 오타 수정

209512 committed:
💡 chore(lecture): 오타 수정

top_n 확인 추가, setting.DEBUG 분기 제거 및 오타 수정

209512 committed:
♻️ refactor(lecture): ALS 추천 시스템 전반의 안정성, 성능, 및 유지보수성 개선 / lecture - tasks, signals.py 추가

data_loader.py
DB 쿼리 성능 최적화:
* _load_user_interactions 내 GroupMember 쿼리에 .select_related("study_group") 추가하여 N+1 쿼리 방지
* DB 접근 효율성 및 속도 개선
증분 학습 (Partial Fit) 안정성 보장:
* build_user_item_matrix에 신규 사용자 조기 감지 로직 구현
* 신규 사용자 감지 시 즉시 Full Training 트리거 플래그(is_new_user_added=True) 반환
* 불필요한 Partial Fit 연산 및 행렬 재구축 방지
하이브리드 피처 통합 강화:
* 상호작용 점수에 북마크, 시간 감쇠된 스터디/검색 로그 포함
* 선호 카테고리 매칭 및 강의 평점을 아이템 피처로 통합하여 점수 합산
벡터화 및 효율성 개선:
* 대용량 데이터의 시간 감쇠 계산에 NumPy 벡터화 연산 (_get_decay_factor_array) 적용하여 성능 극대화
* 강의 카테고리 맵 로드 시 캐시 키 타입 안전성 확보 및 3단계 폴백 로직 정리
메모리 및 자원 관리:
* 대용량 쿼리 처리에 .iterator()와 BATCH_SIZE 상수 활용하여 메모리 사용량 제어 및 안정성 강화

model_trainer.py
테스트 용이성 개선 :
* __init__에 lock_key: Optional[str] 파라미터 추가
* Redis 락 키를 외부에서 주입 가능하게 변경하여 테스트 환경 격리 및 편의성 확보
원자적 저장 후처리 강화:
* _safe_dump_model의 finally 블록에 백업 파일 최종 삭제(Clean Up) 로직 추가
* 저장 실패 시 백업 파일 영구 잔존 방지 및 저장소 관리 안정성 향상
파일 로드 로직 견고화:
* load_model_and_mappings에 EOFError, pickle.UnpicklingError, PermissionError 등 파일 손상 및 시스템 오류 예외 처리 명시적 추가
* 손상된 모델 파일 로드 시 안전하게 None 반환하여 시스템 크래시 방지
증분 학습 행렬 확장 개선:
* Partial Fit 시 Shape 불일치 처리 로직을 COO Matrix 기반 데이터 재배치 방식으로 변경
* NumPy 슬라이싱 방식 대비 희소 행렬 재구축의 정확성 및 안정성 증대
* 맵핑 인덱스에 따른 데이터 손실 위험 최소화

tasks, add_celery_task, signals
ALS 모델 학습 비동기화 :
* Full Training Task (train_full_model_task):
    * 1시간 소프트 / 1시간 5분 하드 타임아웃 설정.
    * 최대 3회 재시도 및 지수 백오프 적용하여 장시간 학습의 운영 안정성 확보.
* Partial Fit Task:
    * 30분 소프트 / 35분 하드 타임아웃 설정 및 재시도 로직 적용.
    * 증분 학습의 안정적인 백그라운드 처리.
ALS Full Training 주기:
* train-full-model-weekly 작업.
* 매주 일요일 새벽 2시 0분에 실행하도록 설정 (day_of_week="0").
* 목적: Full Training의 높은 시스템 리소스 소모를 고려, 주 1회, 사용량이 적은 새벽 시간에 실행 -> 시스템 부하 최소화, 주간 데이터 변동 반영.
ALS Partial Fit 주기:
* partial-fit-daily 작업.
* 매일 새벽 3시 0분에 실행하도록 설정 (hour="3", minute="0").
* 목적: Full Training이 없는 나머지 기간 동안 일별로 누적된 최신 사용자 상호작용을 모델에 반영, 주간 기반 모델과 실시간성 사이의 균형 유지.
자동 캐시 무효화 (Signals):
* CrawledLecture의 post_save, post_delete 시그널에 리시버 연결.
* 강의 메타데이터 캐시 및 인기 강의 캐시를 자동으로 무효화 처리.
* 효과: DB 변경과 캐시 간의 데이터 정합성 자동 확보, 수동 캐시 관리 제거.

209512 committed:
✅ test(lecture): 강의 추천 테스트코드 변경 및 추가

recommendation_integration
recommender
model_trainer
data_loader
tasks
signals

209512 committed:
✅ test(lecture): 추천 테스트 변경 / 캐시키 프리픽스 통일, 테스트 시 로그 출력 조정

209512 committed:
🐛 fix(lecture): model_trainer.py 로그 레벨 조정

209512 committed:
🐛 fix(lecture): implicit.cpu.als.logger 제거 : githun action test 에러

209512 committed:
✅ test(lecture): 추천 tests 로그 출력 제거, 모델 학습 진행률 바 제거.

209512 committed:
🚚 build(lecture): poetry lock

209512 commented:
### 테스트 시 로그 안 뜨게 테스트마다 수정. 
```
# 모델 학습 진행률 바 안 뜨게 설정
import os

os.environ["TQDM_DISABLE"] = "1"

import sys
from unittest.mock import MagicMock

sys.modules["tqdm"] = MagicMock()
sys.modules["tqdm.auto"] = MagicMock()
```
```
# 로그 출력 레벨 설정
@classmethod
    def setUpClass(cls) -> None:
        """클래스 레벨 설정"""
        super().setUpClass()

        logging.getLogger("apps.lecture.services.recommendation_service.model_trainer").setLevel(logging.CRITICAL + 1)
        logging.getLogger("apps.lecture.services.recommendation_service.recommender").setLevel(logging.CRITICAL + 1)
        logging.getLogger("apps.lecture.services.recommendation_service.data_loader").setLevel(logging.CRITICAL + 1)
        logging.getLogger("apps.lecture.tasks").setLevel(logging.CRITICAL + 1)

        # 외부 라이브러리 경고 억제
        warnings.filterwarnings("ignore", category=RuntimeWarning, module="implicit")
        warnings.filterwarnings("ignore", module="implicit.utils")
```
---

## model_trainer.py의 logger.info -> logger.debug로 일부 수정
```
#  info → debug  
logger.debug(f"[ALS][LOCK] Lock busy. Retry {attempt + 1}/{MAX_CACHE_LOAD_RETRIES} ...")
logger.debug("[ALS][BACKUP] Existing model file backed up.")
logger.debug("[ALS][SAVE] Model saved atomically.")
```

209512 commented:
캐시 키 관련 수정
---
### 캐시 키 격리 메커니즘
#### UUID 기반 PREFIX 사용:

- IsolatedRedisTestClient는 각 테스트 인스턴스마다 고유한 UUID prefix 생성:
    - self.CACHE_PREFIX = f"test_{uuid.uuid4().hex}_"
- Django의 KEY_PREFIX 설정으로 적용되어 모든 캐시 키에 자동으로 prefix 추가 [cache.txt:1138-1145](https://github.com/django/django/blob/ad4a9e0f/docs/topics/cache.txt)

test_model_trainer.py, test_recommender 수정
```
def setUp(self) -> None:  
    super().setUp()  
    # IsolatedRedisTestClient의 CACHE_PREFIX 재사용  
    self.cache_prefix = self.CACHE_PREFIX  
    self.test_lock_key = f"{self.CACHE_PREFIX}{ALS_TRAINING_LOCK_KEY}"
```
---

### tearDown에서 캐시 키 삭제의 안전성
#### 병렬 테스트 중 캐시 키 삭제는 문제없음- 이유:

#### 완전한 키 격리:

- 각 테스트는 고유한 UUID prefix 사용
- 예: 테스트1은 test_a1b2c3d4_key, 테스트2는 test_e5f6g7h8_key 사용
- 두 키는 완전히 다른 키이므로 서로 영향 없음

#### Django의 KEY_PREFIX 동작:

- make_key() 메서드에서 prefix 자동 추가
- 각 테스트의 캐시 작업은 자신의 prefix 네임스페이스 내에서만 동작
- 다른 테스트의 키와 절대 충돌하지 않음

- Django 테스트 패턴:
    - Django 자체 테스트도 PID 기반 prefix 사용하여 병렬 테스트 격리 [tests.py:295-303](https://github.com/django/django/blob/ad4a9e0f/tests/cache/tests.py)

현재 test_model_trainer.py: tearDown
```
def tearDown(self) -> None:  
    super().tearDown()  
    cache.delete_many([  
        self.test_lock_key,  
        *[f"{self.CACHE_PREFIX}{key}" for key in ALS_CACHE_KEYS],  
    ])
```
